static/src/javascripts/projects/common/modules/component.js

```diff
@@ -0,0 +1,366 @@
+     * Check if we should auto update, if so, do so
+     */
+    _autoupdate(): void {
+        const setAutoUpdate = () => {
+            // eslint-disable-next-line no-use-before-define
+            this.t = setTimeout(() => update(), this.updateEvery * 1000);
+        };
+        const update = (): void => {
+            this._fetch()
+                .then(resp => {
+                    this.autoupdate(
+                        bonzo.create(resp[this.responseDataKey])[0]
+                    );
+
+                    if (this.autoupdated) {
+                        setAutoUpdate();
+                    }
+                })
+                .catch(() => {
+                    setAutoUpdate();
+                });
+        };
+
+        if (this.autoupdated) {
+            setAutoUpdate();
+        }
+    }
+
+    /**
+     * This is user to edit this.elem before it's rendered
+     * This will help with the rendering performance that
+     * we would lose if rendered then manipulated
+     */
+    // eslint-disable-next-line class-methods-use-this, no-unused-vars
+    prerender(): void {}
+
+    /**
+     * Once the render / decorate methods have been called
+     * This is where you could do your event binding
+     * This function is made to be overridden
+     */
+    // eslint-disable-next-line class-methods-use-this, no-unused-vars
+    ready(elem: ?HTMLElement): void {}
+
+    /**
+     * Once the render / decorate methods have been called
+     * This is where you could do your error event binding
+     * This function is made to be overridden
+     */
+    // eslint-disable-next-line class-methods-use-this, no-unused-vars
+    error(type: string, message?: string): void {}
+
+    /**
+     * This is called whenever a fetch occurs. This includes
+     * explicit fetch calls and autoupdate.
+     */
+    // eslint-disable-next-line class-methods-use-this, no-unused-vars
+    fetched(resp: Object): void {}
+
+    autoupdate(elem: HTMLElement): void {
+        const oldElem = this.elem;
+        this.elem = elem;
+
+        this._prerender();
+        bonzo(oldElem).replaceWith(this.elem);
+    }
+
+    /**
+     * Once we're done with it, remove event bindings etc
+     */
+    // eslint-disable-next-line class-methods-use-this, no-unused-vars
+    dispose(): void {}
+
+    on(
+        eventType: string,
+        elem: string | ((arg: any) => void),
+        handler?: (arg: any) => void
+    ): Component {
+        if (typeof elem === 'function') {
+            const eventHandler = elem;
+            bean.on(this.elem, eventType, eventHandler.bind(this));
+        } else if (typeof handler === 'function') {
+            const selector = !elem.length ? [elem] : elem;
+            bean.on(this.elem, eventType, selector, handler.bind(this));
+        }
+
+        return this;
+    }
+
+    emit(eventName: string, args?: mixed): void {
+        bean.fire(this.elem, eventName, args);
+    }
+
+    getElem(elemName: string): HTMLElement {
```

Comment:
The return type of this method is being junked in a number of places by downcasting to `any` then back up to `HTMLXXXXElement`. Should we change the return type to account for the many possibilities, e.g.

```
getElem(elemName: string): HTMLElement | HTMLLinkElement | HTMLInputElement
```

And perform refinements where necessary

```js
const commentBody = ((this.getElem('body'): any): HTMLInputElement);

if (
  commentBody instanceof 'HTMLInputElement' &&
  commentBody.value.length === 0
) {
  // ...
}
```
I think this is more type-safe approach than casting to `any`. What do you think?

Reply:
Just spoke asynchronously with @GHaberis about this, and we both came to the conclusion that `getElem`'s return type should be nullable, as it is possible that the selector won't match anything.

Do you think this would be a total drag, with lots of null checks everywhere? Saying that, we should bear in mind that having lots of null checks has dramatically reduced the number of errors on the site.

I'd be happy to discuss this in person tomorrow if you prefer.

Reply:
Just seconding Simon comment above ☝️ . I did wonder looking at this whether it's misleading that the return type was `HTMLElement` when it feasibly could be `null`. If we swapped out `qwery` for vanilla JS `flow` would warn about this. As Simon says making the return type nullable will mean we'll need to wrap references to elems derived from `getElem` in null checks, but this would be safer...but yeah it could be a pain to do this.